File: app/src/main/java/com/github/teranes10/androidutils/utils/VersionUpdateUtil.java
package com.github.teranes10.androidutils.utils;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.widget.Toast;

import androidx.core.content.FileProvider;

import java.io.File;

public class VersionUpdateUtil {
    private static final String TAG = "VersionUpdateUtil";

    public static void installAPK(Activity context, String filePath)  {
        File file = new File(filePath);

        if (file.exists()) {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setDataAndType(uriFromFile(context, new File(filePath)), "application/vnd.android.package-archive");
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);


            // startActivity(intent);
            Thread thread = new Thread(() -> {
                try {
                    context.startActivity(intent);

                } catch (Exception e) {
                    e.printStackTrace();
                }

            });
            thread.start();
            context.finish();

        } else {
            Toast.makeText(context, "Installing another version of myDevice.", Toast.LENGTH_LONG).show();
        }
    }

    private static Uri uriFromFile(Context context, File file) {
        return FileProvider.getUriForFile(context, "au.com.softclient.mydevices.provider", file);
    }
}

File: app/src/main/java/com/github/teranes10/androidutils/utils/VersionUpdateUtil.kt
package com.github.teranes10.androidutils.utils

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import com.github.teranes10.androidutils.extensions.ContextExtensions.getUriForFile
import java.io.File

object VersionUpdateUtil {
    private const val TAG = "VersionUpdateUtil"

    fun Context.openUnknownSourcesSettings() {
        val intent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).apply {
            data = Uri.parse("package:$packageName")
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        startActivity(intent)
    }

    fun Context.installAPK(apkFile: File) {
        if (!packageManager.canRequestPackageInstalls()) {
            openUnknownSourcesSettings()
            return
        }

        val uri = getUriForFile(apkFile)
        val intent = Intent(Intent.ACTION_VIEW).apply {
            setDataAndType(uri, "application/vnd.android.package-archive")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }

        try {
            startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            Log.e(TAG, "No Activity found to handle APK installation.", e)
            Toast.makeText(this, "Unable to install APK", Toast.LENGTH_LONG).show()
        } catch (e: Exception) {
            Log.e(TAG, "Unexpected error installing APK", e)
        }
    }
}
